File: src/ruby/tormenta20/models/condicao.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing a Condição (Status Condition) in Tormenta20.
    #
    # Conditions represent temporary states that affect a character's
    # capabilities during play (e.g., Abalado, Cego, Paralisado).
    #
    # @example Find all conditions
    #   Tormenta20.condicoes.all
    #
    # @example Find conditions of a specific type
    #   Tormenta20.condicoes.by_type("medo")
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "abalado")
    # @!attribute [rw] name
    #   @return [String] Condition name in Portuguese (e.g., "Abalado")
    # @!attribute [rw] description
    #   @return [String, nil] Full description of the condition
    # @!attribute [rw] effects
    #   @return [Array<String>] List of mechanical effects
    # @!attribute [rw] condition_type
    #   @return [String, nil] Category (medo, mental, metabolismo, movimento, veneno, sentidos, cansaco, metamorfose)
    # @!attribute [rw] escalates_to
    #   @return [String, nil] ID of the condition this escalates into when reapplied
    class Condicao < Base
      self.table_name = "condicoes"

      TYPES = %w[
        medo
        mental
        metabolismo
        movimento
        veneno
        sentidos
        cansaco
        metamorfose
      ].freeze

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true

      scope :by_type, ->(t) { where(condition_type: t) }
      scope :medo, -> { where(condition_type: "medo") }
      scope :mental, -> { where(condition_type: "mental") }
      scope :metabolismo, -> { where(condition_type: "metabolismo") }
      scope :movimento, -> { where(condition_type: "movimento") }

      def to_h
        {
          id: id,
          name: name,
          description: description,
          effects: effects,
          condition_type: condition_type,
          escalates_to: escalates_to
        }.compact
      end
    end
  end
end

File: src/ts/models/condicao.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"

const TABLE = "condicoes"

/**
 * Category of a condition, used to group conditions by mechanical theme.
 *
 * `medo` · `mental` · `metabolismo` · `movimento` · `veneno` · `sentidos` · `cansaco` · `metamorfose`
 */
export type CondicaoType = "medo" | "mental" | "metabolismo" | "movimento" | "veneno" | "sentidos" | "cansaco" | "metamorfose"

/**
 * A Condição (status condition) in Tormenta20.
 *
 * @example
 * ```ts
 * Condicao.all()
 * Condicao.medo().all()
 * Condicao.find("abalado")
 *
 * const abalado = Condicao.find("abalado")!
 * abalado.escalatesTo   // "apavorado" (next condition in the fear chain)
 * ```
 */
export class Condicao extends BaseModel {
  get description(): string | null { return (this._row.description as string) ?? null }
  /** JSON array of mechanical effect objects. */
  get effects(): unknown[] { return BaseModel.json(this._row.effects, []) }
  get conditionType(): CondicaoType | null { return (this._row.condition_type as CondicaoType) ?? null }
  /** ID of the more severe condition this escalates to, or `null`. */
  get escalatesTo(): string | null { return (this._row.escalates_to as string) ?? null }

  toH() {
    return {
      id: this.id, name: this.name, description: this.description,
      effects: this.effects, condition_type: this.conditionType, escalates_to: this.escalatesTo,
    }
  }

  static query(): Query<Condicao> { return BaseModel.makeQuery(TABLE, Condicao) }
  static all(): Condicao[] { return Condicao.query().all() }
  static find(id: string): Condicao | null { return Condicao.query().find(id) }
  static first(): Condicao | null { return Condicao.query().first() }
  static last(): Condicao | null { return Condicao.query().last() }
  static count(): number { return Condicao.query().count() }

  static byType(type: CondicaoType): Query<Condicao> { return Condicao.query().where("condition_type = ?", type) }
  static medo(): Query<Condicao> { return Condicao.byType("medo") }
  static mental(): Query<Condicao> { return Condicao.byType("mental") }
  static metabolismo(): Query<Condicao> { return Condicao.byType("metabolismo") }
  static movimento(): Query<Condicao> { return Condicao.byType("movimento") }
}
